sulong/projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/multithreading/LLVMPThreadIntrinsics.java

```diff
@@ -0,0 +1,847 @@
+
+        @CompilerDirectives.CompilationFinal
+        FrameSlot spSlot = null;
+
+        private LLVMLanguage language;
+
+        protected RunNewThreadNode(LLVMLanguage language) {
+            super(language);
+            this.language = language;
+        }
+
+        @Override
+        public Object execute(VirtualFrame frame) {
+            // LLVMStack stack = new LLVMStack(1000); // how big should it really be?
+            // LLVMStack.StackPointer sp = stack.newFrame();
+            LLVMStack.StackPointer sp = language.getContextReference().get().getThreadingStack().getStack().newFrame();
+            if (callNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                functionSlot = frame.getFrameDescriptor().findOrAddFrameSlot("function");
+                argSlot = frame.getFrameDescriptor().findOrAddFrameSlot("arg");
+                spSlot = frame.getFrameDescriptor().findOrAddFrameSlot("sp");
+
+                callNode = getCurrentContext(LLVMLanguage.class).getNodeFactory().createFunctionCall(
+                        new MyArgNode(functionSlot),
+                        new LLVMExpressionNode[] {
+                                new MyArgNode(spSlot), new MyArgNode(argSlot)
+                        },
+                        new FunctionType(PointerType.VOID, new Type[] {}, false),
+                        null
+                );
+            }
+            // copy arguments to frame
+            final Object[] arguments = frame.getArguments();
+            Object function = arguments[0];
+            Object arg = arguments[1];
+            frame.setObject(functionSlot, function);
+            frame.setObject(argSlot, arg);
+            frame.setObject(spSlot, sp);
+
+            callNode.executeGeneric(frame);
+            return null;
+        }
+    }
+
+    @NodeChild(type = LLVMExpressionNode.class)
+    public abstract static class LLVMPThreadExit extends LLVMBuiltin {
+        @Specialization
+        protected int doIntrinsic(VirtualFrame frame, Object retval) {
+            // save return value in context for join calls
+            LLVMLanguage.getLLVMContextReference().get().retValStorage.put(Thread.currentThread().getId(), retval);
+            // stop this thread
+            throw new ControlFlowException();
+        }
+    }
+
+    @NodeChild(type = LLVMExpressionNode.class)
+    @NodeChild(type = LLVMExpressionNode.class)
+    public abstract static class LLVMPThreadJoin extends LLVMBuiltin {
+        @Child LLVMStoreNode storeNode;
+
+        @Specialization
+        protected int doIntrinsic(VirtualFrame frame, Object th, Object threadReturn) {
+            if (storeNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                storeNode = LLVMLanguage.getLLVMContextReference().get().getNodeFactory().createStoreNode(LLVMInteropType.ValueKind.POINTER);
+            }
+            long thLong = (long) th;
```

Comment:
If you expect `th` to be a `long` then that should be the type of the function parameter. The Truffle DSL will take care of casting the value and handling `ClassCastException` for you. The same goes for `ThreadReturn` here and other function parameters in this file.